src/shared/reboot-util.c

```diff
@@ -107,3 +108,43 @@ int shall_restore_state(void) {
 
         return r > 0 ? ret : true;
 }
+
+static int xen_kexec_loaded(void) {
+#if HAVE_XENCTRL
+        xc_interface *xch;
+        int r;
+
+        r = dlopen_xenctrl();
+        if (r < 0)
+                return r;
+
+        if (access("/proc/xen", F_OK) < 0)
+                return log_debug_errno(errno, "Cannot access /proc/xen: %m");
+
+        xch = sym_xc_interface_open(NULL, NULL, 0);
+        if (!xch)
+                return log_debug_errno(errno, "Failed to open Xen control interface: %m");
+
+        r = sym_xc_kexec_status(xch, KEXEC_TYPE_DEFAULT);
+
+        sym_xc_interface_close(xch);
+
+        return r;
+#else
+        return -EOPNOTSUPP;
+#endif
+}
+
+bool kexec_loaded(void) {
+       _cleanup_free_ char *s = NULL;
+       int r;
+
+       r = xen_kexec_loaded();
+       if (r >= 0)
+               return r != 0;
```

Comment:
`xen_kexec_loaded()` is a boolean function with negative errno. Hence, in this case, please simply use `return r;`.